### PR TITLE
Prevent deferred File.Close and wrap errors while putting CRDs into xpls provider package cache

### PR DESCRIPTION
**Pull request by ulucinar**
### Description of your changes

Removes the deferred `File.Close` call in `cache.entry.writeObjects` to decrease the # of simultaneously open files and wraps the error returned from `cache.entry.appendToPackageJSON` to provide more context.

I have:

- [x] Read and followed Upbound's [contribution process](https://git.io/fj2m9).
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR, as appropriate.

### How has this code been tested

Tested with Go 1.18.10 on a test configuration package with the following command:
```shell
/tmp/up xpkg dep xpkg.upbound.io/upbound/provider-gcp
```
(I've initially observed this issue with `up@v.17.0`, which uses Go `1.18`)

### CODE REVIEW

File: internal/xpkg/dep/cache/entry.go [hunk 1]
```diff
@@ -246,23 +246,12 @@ func (e *entry) writeObjects(objs []runtime.Object) (*flushstats, error) { // no
...
-
-		fb, err := f.Write(yb)
-		if err != nil {
-			return stats, err
-		}
-
-		if fb == 0 {
-			return stats, errors.New(errFailedToCreateCRD)
+		if err := os.WriteFile(filepath.Join(e.location(), fmt.Sprintf(crdNameFmt, name)), yb, 0o600); err != nil {
+			return stats, errors.Wrap(err, errFailedToCreateCRD)
```
Comment (sttts) [inline internal/xpkg/dep/cache/entry.go:250]: Have you double checked that the filename is part of the errors from `os.WriteFile`?

File: internal/xpkg/dep/cache/entry.go [hunk 2]
```diff
@@ -246,23 +246,12 @@ func (e *entry) writeObjects(objs []runtime.Object) (*flushstats, error) { // no
...
 		}
 
 		if err := e.appendToPackageJSON(jb); err != nil {
-			return stats, errors.New(errFailedToCreateCRD)
+			return stats, errors.Wrap(err, errFailedToCreateCRD)
```
Comment (sttts) [inline internal/xpkg/dep/cache/entry.go:254]: that error is `failed to create crd`. Can a user understand what it means?

File: internal/xpkg/dep/cache/entry.go [hunk 3]
```diff
@@ -246,23 +246,12 @@ func (e *entry) writeObjects(objs []runtime.Object) (*flushstats, error) { // no
...
-		defer f.Close() // nolint:errcheck
-
-		fb, err := f.Write(yb)
-		if err != nil {
-			return stats, err
-		}
-
-		if fb == 0 {
-			return stats, errors.New(errFailedToCreateCRD)
+		if err := os.WriteFile(filepath.Join(e.location(), fmt.Sprintf(crdNameFmt, name)), yb, 0o600); err != nil {
```
Comment (tnthornton) [inline internal/xpkg/dep/cache/entry.go:249]: I think `os.Writefile` is going to side step the use of `afero.Fs` here. We can fix that up by using afero's API xref https://github.com/spf13/afero#using-aferos-utility-functions. So something like this:

```suggestion
		if err := afero.WriteFile(e.fs, filepath.Join(e.location(), fmt.Sprintf(crdNameFmt, name)), yb, 0o600); err != nil {
```
may work. Here's another example of doing that in this repo https://github.com/upbound/up/blob/main/internal/xpkg/build_test.go#L171.

File: internal/xpkg/dep/cache/entry.go [hunk 4]
```diff
@@ -36,10 +36,11 @@ const (
 	crdNameFmt = "%s.yaml"
 	delim      = "\n"
 
-	errFailedToCreateMeta      = "failed to create meta file in entry"
-	errFailedToCreateImageMeta = "faile to create image meta entry"
-	errFailedToCreateCRD       = "failed to create crd"
-	errNoObjectsToFlushToDisk  = "no objects to flush"
+	errFailedToCreateMeta       = "failed to create meta file in entry"
+	errFailedToCreateImageMeta  = "failed to create image meta entry"
+	errFailedToCreateCacheEntry = "failed to create a cache entry for the CRD at path: %s"
```
Comment (sttts) [inline internal/xpkg/dep/cache/entry.go:41]: nit: I believe to have seen a `Fmt` suffix in most places if it is a format.

File: internal/xpkg/dep/cache/entry.go [hunk 5]
```diff
@@ -36,10 +36,11 @@ const (
...
+	errFailedToAppendCRD        = "failed to add the CRD to the package: %s"
```
Comment (sttts) [inline internal/xpkg/dep/cache/entry.go:42]: this won't look good with `errors.Wrapf`:  `failed to add the CRD to the package: package-name: some error`. Would rather use:
```suggestion
	errFailedToAppendCRD        = "failed to add the CRD to package %q"
```

File: internal/xpkg/dep/cache/entry.go (hunk 1 above)
Comment (ulucinar) [reply internal/xpkg/dep/cache/entry.go]: Hi @sttts,
I think the file being processed is not always part of the error message coming from `os.WriteFile`. In my case, for instance, it was just `too many open files`. Added the cache entry's path into the error's context.

File: internal/xpkg/dep/cache/entry.go (hunk 2 above)
Comment (ulucinar) [reply internal/xpkg/dep/cache/entry.go]: Done. I've replaced that error message as `failed to create a cache entry for the CRD at path: %s`. I did not include the CRD name in the error message as the filename of the cache entry already has that information.

File: internal/xpkg/dep/cache/entry.go (hunk 3 above)
Comment (ulucinar) [reply internal/xpkg/dep/cache/entry.go]: Hi @tnthornton,
Thanks, done.

File: internal/xpkg/dep/cache/entry.go (hunk 5 above)
Comment (ulucinar) [reply internal/xpkg/dep/cache/entry.go]: Done.

**Review by tnthornton — APPROVED** [review]
Thanks @ulucinar ! My concern was addressed, so LGTM 👍

**Pull request merged**